### PR TITLE
Websocketfactory

### PR DESCRIPTION
**Pull request by Trolldemorted**
What do you think?

cc @golf1052 @mitchcapper

### CODE REVIEW

**Comment by mitchcapper** [comment]
looks fine although you are not touching PushServiceSocket at all.   Also there is no "Default" or provided implementer of it not the end of the world but just a bit harder for someone to start using the lib.  This will need an aux patch to support the server certificate in PushServiceSocket.

If you wanted to push this I could create a new PR to add a default implementation and pushservicesocket changes.

**Comment by Trolldemorted** [comment]
>looks fine although you are not touching PushServiceSocket at all.

not *yet* :) Doesn't 31a12eaa290e0a94bfe74447f0ed5796cd9c76e2 work on every platform except uwp (which has its own workaround)?

Yeah a default implementor would be nice, right now this is the status quo of my ~weekend~ midweek hack. Question is what would be the best location for the default/example impl? The test project targets netcoreapp2.0, how about that?

**Comment by mitchcapper** [comment]
31a12ea does it just is horrid security practice especially in a secure messaging app:)  Extend it somehow to take the certificate to check against like #17 and it should be good.  

I would provide a default handler right in the library or not at all.  If people want an example they can just use Signal-Windows.

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs [hunk 1]
```diff
@@ -106,25 +99,17 @@ public WebSocketRequestMessage ReadRequestBlocking()
...
         {
-            Tuple<CountdownEvent, uint, string> t = new Tuple<CountdownEvent, uint, string>(new CountdownEvent(1), 0, null);
             WebSocketMessage message = new WebSocketMessage
             {
                 Type = WebSocketMessage.Types.Type.Request,
                 Request = request
             };
-            OutgoingRequests.AddOrUpdate(request.Id, t, (k, v) => t);
-            WebSocket.OutgoingQueue.Add(message.ToByteArray());
-            return await Task.Run(() =>
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs:119]: Very glad this Task.Run is gone

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs [hunk 2]
```diff
@@ -138,11 +123,13 @@ public void SendResponse(WebSocketResponseMessage response)
                 Type = WebSocketMessage.Types.Type.Response,
                 Response = response
             };
-            WebSocket.OutgoingQueue.Add(message.ToByteArray());
+            SignalWebSocket.SendMessage(message.ToByteArray());
         }
 
         private void SendKeepAlive(CancellationToken token, object state)
         {
+            throw new NotImplementedException();
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs:138]: Does this ever get called?

File: libsignal-service-dotnet/websocket/ISignalWebSocket.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace libsignalservicedotnet.websocket
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/ISignalWebSocket.cs]: namespace should be libsignalservice.websocket

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs [hunk 4]
```diff
@@ -45,49 +47,40 @@ internal SignalWebSocketConnection(CancellationToken token, string httpUri, Cred
...
+            var msg = WebSocketMessage.Parser.ParseFrom(e.Message);
             if (msg.Type == WebSocketMessage.Types.Type.Request)
             {
-                Logger.LogTrace("Adding request to IncomingRequests");
                 IncomingRequests.Add(msg.Request);
             }
             else if (msg.Type == WebSocketMessage.Types.Type.Response)
             {
-                Logger.LogTrace("Adding response {0} ({1} {2})", msg.Response.Id, msg.Response.Status, Encoding.UTF8.GetString(msg.Response.Body.ToByteArray()));
                 var t = new Tuple<CountdownEvent, uint, string>(null, msg.Response.Status, Encoding.UTF8.GetString(msg.Response.Body.ToByteArray()));
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs]: This line should be removed.

File: libsignal-service-dotnet/websocket/ISignalWebSocket.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace libsignalservicedotnet.websocket
+{
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+    public class SignalWebSocketMessageReceivedEventArgs
+    {
+        public Stream Message { get; set; }
+    }
+
+    public interface ISignalWebSocket : IDisposable
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/ISignalWebSocket.cs]: What is actually implementing this?

File: libsignal-service-dotnet/SignalServiceMessagePipe.cs [hunk 6]
```diff
@@ -99,18 +103,17 @@ internal SignalServiceMessagePipe(CancellationToken token, SignalWebSocketConnec
                 Body = ByteString.CopyFrom(Encoding.UTF8.GetBytes(JsonUtil.ToJson(list)))
             };
             requestmessage.Headers.Add("content-type:application/json");
-            Logger.LogDebug("Sending message {0}", requestmessage.Id);
-            var t = Websocket.SendRequest(requestmessage);
-            await t;
-            if (t.IsCompleted)
+            var sendTask = Websocket.SendRequest(requestmessage).Task;
+            if (await Task.WhenAny(sendTask, Task.Delay(10*1000)) == sendTask)
```
Comment (golf1052) [inline libsignal-service-dotnet/SignalServiceMessagePipe.cs]: You should add a cancellation token to the Task.Delay so that it gets cancelled if the request is successful and we don't have a bunch of timers laying around: https://stackoverflow.com/a/22078975/6681022

File: libsignal-service-dotnet/SignalServiceMessagePipe.cs [hunk 7]
```diff
@@ -134,16 +137,15 @@ internal SignalServiceMessagePipe(CancellationToken token, SignalWebSocketConnec
                 Path = $"/v1/profile/{address.E164number}"
             };
 
-            var t = Websocket.SendRequest(requestMessage);
-            await t;
-            if (t.IsCompleted)
+            var sendTask = Websocket.SendRequest(requestMessage).Task;
+            if (await Task.WhenAny(sendTask, Task.Delay(10 * 1000)) == sendTask)
```
Comment (golf1052) [inline libsignal-service-dotnet/SignalServiceMessagePipe.cs]: Same as above.

File: libsignal-service-dotnet/libsignal-service-dotnet.csproj [hunk 8]
```diff
@@ -33,7 +33,6 @@
     <PackageReference Include="Microsoft.Extensions.Logging" Version="1.1.2" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.1" />
     <PackageReference Include="Portable.BouncyCastle" Version="1.8.2" />
-    <PackageReference Include="System.Net.WebSockets.Client" Version="4.3.2" />
```
Comment (golf1052) [inline libsignal-service-dotnet/libsignal-service-dotnet.csproj:36]: Why was this removed?

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs [hunk 9]
```diff
@@ -106,25 +99,17 @@ public WebSocketRequestMessage ReadRequestBlocking()
...
-                if (t.Item1.Wait(10 * 1000, Token))
-                {
-                    var handledTuple = OutgoingRequests[request.Id];
-                    return new Tuple<uint, string>(handledTuple.Item2, handledTuple.Item3);
-                }
-                throw new IOException("wait for confirmation timeout");
-            });
+            var messageSendResult = new TaskCompletionSource<(uint, string)>();
+            OutgoingRequests.AddOrUpdate(request.Id, messageSendResult, (k, v) => messageSendResult);
+            SignalWebSocket.SendMessage(message.ToByteArray());
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs]: await here?

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs [hunk 10]
```diff
@@ -138,11 +123,13 @@ public void SendResponse(WebSocketResponseMessage response)
                 Type = WebSocketMessage.Types.Type.Response,
                 Response = response
             };
-            WebSocket.OutgoingQueue.Add(message.ToByteArray());
+            SignalWebSocket.SendMessage(message.ToByteArray());
```
Comment (golf1052) [inline libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs:133]: await here?

File: libsignal-service-dotnet/push/ProvisioningSocket.cs [hunk 11]
```diff
@@ -1,52 +1,55 @@
...
+        private readonly string WsUri;
+        private readonly BlockingCollection<Stream> IncomingRequests = new BlockingCollection<Stream>(new ConcurrentQueue<Stream>());
+
+        public ProvisioningSocket(string httpUri, ISignalWebSocketFactory webSocketFactory)
+        {
+            WsUri = httpUri.Replace("https://", "wss://")
+                .Replace("http://", "ws://") + "/v1/websocket/provisioning/";
+            SignalWebSocket = webSocketFactory.CreateSignalWebSocket();
+            SignalWebSocket.MessageReceived += SignalWebSocket_MessageReceived;
+            SignalWebSocket.ConnectAsync(new Uri(WsUri));
```
Comment (golf1052) [inline libsignal-service-dotnet/push/ProvisioningSocket.cs]: Shouldn't connect in the constructor since this needs an await

File: libsignal-service-dotnet/websocket/ISignalWebSocket.cs (hunk 5 above)
Comment (mitchcapper) [reply libsignal-service-dotnet/websocket/ISignalWebSocket.cs]: @golf1052  Every app that uses this lib will have to.  Signal-Windows will have a reference implementation.  I was commenting above about the possibility of putting a default implementation in the lib (although to make it .net core relevant we would still need the dual build package).

File: libsignal-service-dotnet/libsignal-service-dotnet.csproj (hunk 8 above)
Comment (Trolldemorted) [reply libsignal-service-dotnet/libsignal-service-dotnet.csproj:36]: Because we don't need it anymore!

File: libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs (hunk 2 above)
Comment (Trolldemorted) [reply libsignal-service-dotnet/websocket/SignalWebSocketConnection.cs:138]: Not at the moment, but apparently libsignal-service-java sends these. Might want to send them too in the future?

**Comment by Trolldemorted** [comment]
Yeah we should pin the cert, or add an IHTTPClient interface to our interface.

Can we provide a default handler in the library? The lib is still netstandard1.4, is there a way to get the certificate validation on some/most platforms?

**Comment by mitchcapper** [comment]
Well we can create a library targeting netstandard1.4 and netapp 2.1 (replaced by netstandard 2.1 when that comes out) that would offer a default handler that had certificate validation for most platforms.

**Comment by Trolldemorted** [comment]
How can we target two targets? Or should we release two binaries?

**Comment by mitchcapper** [comment]
PR #17  already does it.  nuget and the project file both natively  support multi-targetting.  they will automatically use the highest target (so everything will use .net 1.4 minus netapp 2.1 (core 2.1) apps which will take the alternate binary in the nuget package.   It creates one nuget file with both.

**Comment by Trolldemorted** [comment]
Merging this now (because the identity key exception fix has to be deployed asap), let's continue the discussion in #9
